impl/src/main/java/org/ehcache/impl/internal/classes/ClassInstanceProvider.java

```diff
@@ -49,6 +51,7 @@
    * Instances provided by this provider vs their counts.
    */
   protected final ConcurrentWeakIdentityHashMap<T, AtomicInteger> providedVsCount = new ConcurrentWeakIdentityHashMap<T, AtomicInteger>();
+  protected final Set<T> instantiated = new HashSet<T>();
```

Comment:
And one more advantage of `ConcurrentWeakIdentityHashMap` is that you are not holding a strong reference to the instance.